torchtitan/torchtitan/parallelisms/pipelining_utils.py

```diff
@@ -0,0 +1,114 @@
+# LICENSE file in the root directory of this source tree.
+import os
+from typing import Tuple
+
+from torch.distributed.pipelining.schedules import (
+    get_schedule_class,
+    PipelineScheduleMulti,
+    PipelineScheduleSingle,
+)
+from torchtitan.logging import logger
+
+
+def generate_split_points(job_config, pp_dim, model_config):
+    schedule_class = get_schedule_class(
+        job_config.experimental.pipeline_parallel_schedule
+    )
+    if issubclass(schedule_class, PipelineScheduleSingle):
+        num_stages_per_rank = 1
+    elif issubclass(schedule_class, PipelineScheduleMulti):
+        # Multi-stage schedules support more than 2 stages per rank, but this is the default if
+        # no pipeline split is specified
+        num_stages_per_rank = 2
+    else:
+        raise ValueError(
+            f"Unsupported pipeline schedule: {job_config.experimental.pipeline_parallel_schedule}"
+        )
+    total_stages = pp_dim * num_stages_per_rank
+    num_layers = model_config.n_layers
+    if total_stages > num_layers:
+        raise ValueError("Total stages cannot be greater than the number of layers")
+
+    base_interval = num_layers // total_stages
+    extra_layers = num_layers % total_stages
+
+    splits = []
+    current_layer = 0
+    for i in range(total_stages - 1):
+        if i == 0:
+            current_layer += base_interval
+        else:
+            # Middle stages get an extra layer if there are any remaining
+            if extra_layers > 0:
+                current_layer += base_interval + 1
+                extra_layers -= 1
+            else:
+                current_layer += base_interval
+        splits.append("layers." + str(current_layer))
+    logger.info(
+        f"No 'pipeline_parallel_split_points' so the generated splits are: {splits} \
+This may be sub-optimal as the number of layers per stage may be unbalanced."
+    )
+    return splits
+
+
+def build_pipeline_schedule(job_config, stages, loss_fn):
+    pp_schedule_csv = job_config.experimental.pipeline_parallel_schedule_csv
+
+    # Validate that pp_schedule_csv is a valid path
+    if pp_schedule_csv and not os.path.isfile(pp_schedule_csv):
+        raise FileNotFoundError(
+            f"The specified path {pp_schedule_csv} does not exist or is not a file."
+        )
+
+    schedule_class = get_schedule_class(
+        job_config.experimental.pipeline_parallel_schedule
+    )
+
+    looped_schedule = issubclass(schedule_class, PipelineScheduleMulti)
+    logger.info(
+        f"Using pipeline schedule {job_config.experimental.pipeline_parallel_schedule}"
+    )
+    n_microbatches = job_config.experimental.pipeline_parallel_microbatches
+    if n_microbatches is None:
+        n_microbatches = job_config.experimental.pipeline_parallel_degree
+
+    schedule = schedule_class(
+        stages if looped_schedule else stages[0],
+        n_microbatches=n_microbatches,
+        loss_fn=loss_fn,
+    )
```

Comment:
stages랑 n_microbatches, loss_fn 넣고 call
안에서 무슨 일이 일어나나 하나만 까보면 -- https://github.com/pytorch/pytorch/blob/4959784dac633cc23fc59401eb8cfcae47a07f4c/torch/distributed/pipelining/schedules.py#L549
[_step_microbatches](https://github.com/pytorch/pytorch/blob/4959784dac633cc23fc59401eb8cfcae47a07f4c/torch/distributed/pipelining/schedules.py#L555-L561)란 함수 내에 microbatch에 대해 loss구하고 p2p 통신하는 것 까지 구현되어 있음.
```python
    def _step_microbatches(
        self,
        arg_mbs: Optional[List] = None,
        kwarg_mbs: Optional[List] = None,
        target_mbs: Optional[List] = None,
        losses: Optional[List] = None,
    ):
        """
        Run one iteration of the pipeline schedule with list of microbatches.
        Will go through all the microbatches according to the GPipe schedule.

        Args:
            microbatches: list of microbatch args.
        """
        arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs, losses)

        if not self._stage_initialized:
            self._initialize_stage(arg_mbs[0], kwarg_mbs[0])

        # Delay send waits
        fwd_sends_to_wait: List[dist.Work] = []

        # Run microbatches
        for i in range(self._n_microbatches):
            with record_function(f"Forward {i}"):
                ops = self._stage.get_fwd_recv_ops(i)
                works = _sorted_batch_p2p(ops, desc="fwd_recv")
                for work in works.values():
                    work.wait()

                output = self._stage.forward_one_chunk(i, arg_mbs[i], kwarg_mbs[i])  # type: ignore[index]

                ops = self._stage.get_fwd_send_ops(i)
                works = _sorted_batch_p2p(ops, desc="fwd_send")
                fwd_sends_to_wait.extend(works.values())

            logger.debug("[%s] Forwarded microbatch %s", self._stage.stage_index, i)

            self._maybe_compute_loss(self._stage, output, target_mbs, i)

        # Wait for all forward sends to finish
        # This should not have performance impact because by the time the first
        # backward arrives all the forward sends should have been finished.
        for work in fwd_sends_to_wait:
            work.wait()

        # No loss function, no need to run backward
        if not self._has_backward:
            return

        # Run backward
        # Delay send waits
        bwd_sends_to_wait: List[dist.Work] = []
        for i in range(self._n_microbatches):
            with record_function(f"Backward {i}"):
                ops = self._stage.get_bwd_recv_ops(i)
                works = _sorted_batch_p2p(ops, desc="bwd_recv")
                for work in works.values():
                    work.wait()

                loss = self._maybe_get_loss(self._stage, i)
                self._stage.backward_one_chunk(
                    i, loss=loss, last_backward=i == self._n_microbatches - 1
                )

                ops = self._stage.get_bwd_send_ops(i)
                works = _sorted_batch_p2p(ops, desc="bwd_send")
                bwd_sends_to_wait.extend(works.values())

            logger.debug("[%s] Backwarded microbatch %s", self._stage.stage_index, i)

        # Return losses if there is a container passed in
        self._update_losses(self._stage, losses)

        # Wait for all backward sends to finish
        for work in bwd_sends_to_wait:
            work.wait()
```